autopilot/hostfilter.go

```diff
@@ -46,6 +50,11 @@ func isUsableHost(cfg api.AutopilotConfig, gs api.GougingSettings, rs api.Redund
 // isUsableContract returns whether the given contract is usable and whether it
 // can be renewed, along with a list of reasons why it was deemed unusable.
 func isUsableContract(cfg api.AutopilotConfig, h Host, c api.Contract, bh uint64) (usable bool, refresh bool, renew bool, reasons []string) {
+	if isOutOfCollateral(cfg, h, c) {
+		reasons = append(reasons, "out of collateral")
+		renew = false
+		refresh = true
```

Comment:
I think if we do this we also need some sort of protection against hosts which simply don't have any collateral left to give in addition to the one already in the contract upon refresh.

Right now it's very simple for a host to force you into a refresh loop. which allows the host to profit off of the contract cost.

Another thing to consider is updating `PrepareContractRenewal` to match this logic. e.g. we could change it so that we only ever refresh if the refresh gets us above the collateral threshold.

Reply:
I added a `TODO` that covers the first part, we should analyse the errors we receive from the hosts when trying to refresh/renew contracts and give up after a certain amount of attempts or depending on the error just right away. I tried avoiding the refresh loop by checking whether the contract is deemed unusable due to the collateral and ignoring the contract when the updated collateral would cause a loop or exceed the host's max collateral.

I did not update `PrepareContractRenewal` but copied over the `bus`'s way of calculating the collateral. Perhaps this is another reason to move this out of the `bus` and have the collateral be a parameter that has to be supplied 🤔  (something we discussed couple of weeks ago)

Reply:
Yeah we can move the collateral out of the bus to match contract formation. I think compared to a couple of weeks ago it's pretty obvious now that being consistent with contract formations has its merits for figuring out whether to renew at all.